### PR TITLE
chore(sql): time interval intrinsics for parquet partitions

### PR DESCRIPTION
**Pull request by puzpuzpuz**
Includes changes from #5015

### CODE REVIEW

File: core/rust/qdbr/src/parquet_read/decode.rs [hunk 1]
```diff
@@ -333,6 +333,82 @@ impl<R: Read + Seek> ParquetDecoder<R> {
...
+                // left boundary
+                if timestamp == min_value && min_value != max_value {
+                    // have to decode the group and search in it
+                    return Ok(2 * (row_group_idx + 1) as u64);
+                }
+                // to the left of the row group
+                if timestamp < min_value {
+                    if matched_prev_max {
+                        // right boundary of the previous row group
+                        return Ok((2 * row_group_idx + 1) as u64);
```
[the remark this answers is not in the document]
Comment (puzpuzpuz) [reply core/rust/qdbr/src/parquet_read/decode.rs]: Simplified the documented the code in 0725ef115d037a6dba534edace5efe19fb0266ae

File: core/rust/qdbr/src/parquet_read/decode.rs [hunk 2]
```diff
@@ -333,6 +333,82 @@ impl<R: Read + Seek> ParquetDecoder<R> {
         }
         Ok(())
     }
+
+    pub fn find_row_group_by_timestamp(
+        &mut self,
```
[the remark this answers is not in the document]
Comment (puzpuzpuz) [reply core/rust/qdbr/src/parquet_read/decode.rs]: Removed from this method and `read_column_chunk_stats` in 5d64ab45890c21c4439070aa6aef8947702fdeb4

File: core/src/main/java/io/questdb/cairo/TableReader.java [hunk 3]
```diff
@@ -954,12 +985,11 @@ private long openPartition0(int partitionIndex) {
...
                         openPartitionInfo.setQuick(offset + PARTITIONS_SLOT_OFFSET_FORMAT, PartitionFormat.PARQUET);
 
-                        // TODO(puzpuzpuz): we need to read txn's Parquet file size and use it when reading
-                        //                  that's because O3 appends new versions of the file to its end
-                        long fd = TableUtils.openRO(ff, path.$(), LOG);
-                        assert fd > 0;
+                        final long fd = TableUtils.openRO(ff, path.$(), LOG);
                         assert openPartitionInfo.getQuick(offset + PARTITIONS_SLOT_OFFSET_PARQUET_FD) == -1;
                         openPartitionInfo.setQuick(offset + PARTITIONS_SLOT_OFFSET_PARQUET_FD, fd);
+                        final long fileSize = this.txFile.getPartitionParquetFileSize(partitionIndex);
```
[the remark this answers is not in the document]
Comment (puzpuzpuz) [reply core/src/main/java/io/questdb/cairo/TableReader.java]: Added in 5e5f311ee07496e8d4504e1ad4bcea33dbbb528d

File: core/src/main/java/io/questdb/cairo/sql/PartitionFrame.java [hunk 4]
```diff
@@ -24,38 +24,26 @@
...
-     */
-    int getParquetRowGroup();
-
-    /**
-     * @return first row index of the row group within Parquet partition;
-     * shouldn't be called for native partitions
+     * @return parquet decoder initialized for the partition for parquet partitions; null for native partitions
      */
-    int getParquetRowGroupLo();
+    PartitionDecoder getParquetDecoder();
```
[the remark this answers is not in the document]
Comment (puzpuzpuz) [reply core/src/main/java/io/questdb/cairo/sql/PartitionFrame.java:48]: Good idea. Added in 5e5f311ee07496e8d4504e1ad4bcea33dbbb528d

File: core/rust/qdbr/src/parquet_read/jni.rs [hunk 5]
```diff
@@ -141,6 +144,34 @@ pub extern "system" fn Java_io_questdb_griffin_engine_table_parquet_PartitionDec
     }
 }
 
+// See PartitionDecoder for more info on the returned value format.
+#[no_mangle]
+pub extern "system" fn Java_io_questdb_griffin_engine_table_parquet_PartitionDecoder_findRowGroupByTimestamp(
+    mut env: JNIEnv,
+    _class: JClass,
+    decoder: *mut ParquetDecoder<NonOwningFile>,
```
[the remark this answers is not in the document]
Comment (puzpuzpuz) [reply core/rust/qdbr/src/parquet_read/jni.rs]: Changed here and for the read stats function in 5e5f311ee07496e8d4504e1ad4bcea33dbbb528d

File: core/rust/qdbr/src/parquet_read/decode.rs [hunk 6]
```diff
@@ -333,6 +333,82 @@ impl<R: Read + Seek> ParquetDecoder<R> {
...
+        timestamp: i64,
+        row_lo: usize,
+        row_hi: usize,
+        timestamp_column_index: u32,
+    ) -> ParquetResult<u64> {
+        let timestamp_column_index = timestamp_column_index as usize;
+        let column_type = self.columns[timestamp_column_index].column_type;
```
[the remark this answers is not in the document]
Comment (puzpuzpuz) [reply core/rust/qdbr/src/parquet_read/decode.rs:354]: Added the validation and fixed the one from read stats method in 5e5f311ee07496e8d4504e1ad4bcea33dbbb528d

File: core/rust/qdbr/parquet2/src/read/metadata.rs [hunk 7]
```diff
@@ -60,21 +60,28 @@ pub fn read_metadata_with_size<R: Read + Seek>(
...
-    // reader.seek(SeekFrom::End(-(default_end_len as i64)))?;
 
     let mut buffer = Vec::with_capacity(default_end_len);
-    reader
+    let read_bytes = reader
         .by_ref()
         .take(default_end_len as u64)
         .read_to_end(&mut buffer)?;
+    if read_bytes < default_end_len {
+        return Err(Error::oos(
```
[the remark this answers is not in the document]
Comment (puzpuzpuz) [reply core/rust/qdbr/parquet2/src/read/metadata.rs:70]: `read_to_end` may read less bytes than requested. If that's the case, we'll panic at the `buffer[default_end_len - 4..] != PARQUET_MAGIC` line. This is a rather defensive check, so I don't mind removing it. WDYT?

[the remark this answers is not in the document]
Comment (puzpuzpuz) [reply core/rust/qdbr/parquet2/src/read/metadata.rs]: Removed in 0157b36650b31f5598a93036b9fdb846eacd95bf

**Review by eugenels — APPROVED** [review]

**Comment by puzpuzpuz** [comment]
@eugenels thanks for the review!

**Pull request merged**